### PR TITLE
WIP(@mkhatib): Add amp-form extension for handling and polyfilling forms.

### PR DESCRIPTION
**Pull request by mkhatib**
@dvoytenko can you take a look at the `amp-form.js` and `examples/forms.amp.html` (You can try [the behavior here](https://amp-mk-1.herokuapp.com/examples.build/forms.amp.max.html)) and see if this is the right approach to continue with this?

I've highlighted some of the stuff that we need to get support for in #3343. I didn't want to implement all of them in one PR, so let me know if what I have here is good for a first PR and I can write some tests and fix it up some more.

cc/ @cramforce

### CODE REVIEW

**Comment by mkhatib** [comment]
@dvoytenko let me know if this looks good in general so I can proceed with adding tests.

File: examples/forms.amp.html [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    <script async custom-element="amp-form" src="https://cdn.ampproject.org/v0/amp-form-0.1.js"></script>
+    <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.1.js"></script>
+    <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
+    <script async src="https://cdn.ampproject.org/v0.js"></script>
+    <style amp-custom>
+        form.success {
```
Comment (dvoytenko) [inline examples/forms.amp.html]: See doc comments on class names. I believe it should be instead `amp-submit-success` or even `amp-form-submit-success`.

File: examples/forms.amp.html [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            border: 1px solid #0e5829;
+        }
+
+        form.success .amp-form-messages {
```
Comment (dvoytenko) [inline examples/forms.amp.html]: Do we really need `.amp-form-messages`?

File: examples/forms.amp.html [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            color: #0e5829;
+        }
+
+        form.success fieldset,
```
Comment (dvoytenko) [inline examples/forms.amp.html]: This looks like should be part of AMP's default CSS with an option to override.

File: examples/forms.amp.html [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        form.error {
+            border: 1px solid #dc4e41;
+        }
+
+        form.error .amp-form-messages {
+            color: #dc4e41;
+        }
+
+        form.success .success-message {
+            display: block;
```
Comment (dvoytenko) [inline examples/forms.amp.html]: IMHO it's better to have a cancel-display-none structure for CSS so that we don't need to worry about other display types, such as flexbox.

File: examples/forms.amp.html [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        form .now-valid-message {
+            display: none;
+        }
+
+        form.invalid .invalid-message {
+            display: block;
+            color: #dc4e41;
+        }
+
+        /* was invalid but now valid */
```
Comment (dvoytenko) [inline examples/forms.amp.html]: I don't understand this. Could you please clarify?

File: extensions/amp-form/0.1/amp-form.js [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+import {toArray} from '../../../src/types';
+import {onDocumentReady} from '../../../src/document-ready';
+import {isExperimentOn} from '../../../src/experiments';
+
+/** @const @type {string} */
+const TAG = 'amp-form';
+
+class AmpForm {
+
+  constructor(element, win) {
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: jsdoc please

File: extensions/amp-form/0.1/amp-form.js [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    /** @const @private {!Window} */
+    this.win_ = win;
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: you don't need to pass `win` here since you can get it from `element.ownerDocument.defaultView`

File: extensions/amp-form/0.1/amp-form.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    this.form_ = element;
+
+    /** @const @private {!Templates} */
+    this.templates_ = templatesFor(this.win_);
+
+    /** @const @private {!Xhr} */
+    this.xhr_ = xhrFor(this.win_);
+
+    /** @const @private {string} */
+    this.method_ = this.form_.getAttribute('method') || 'POST';
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: Default to `GET`

File: extensions/amp-form/0.1/amp-form.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    /** @const @private {string} */
+    this.action_ = user.assert(this.form_.getAttribute('action'),
+        'amp-form requires action attribute %s', this.form_);
+
+    this.installSubmitHandler_();
+  }
+
+  /** @private */
+  installSubmitHandler_() {
+    this.form_.addEventListener('submit', e => {
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: So, all/most of these could go to core for "submit" event capture.

File: package.json [hunk 10]
```diff
@@ -41,6 +41,7 @@
     "eslint": "1.10.3",
     "eslint-plugin-google-camelcase": "0.0.2",
     "finalhandler": "0.4.0",
+    "formidable": "^1.0.17",
```
Comment (erwinmombay) [inline package.json]: remove ^

File: extensions/amp-form/0.1/amp-form.js [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        el => el.tagName == 'TEMPLATE' && el.hasAttribute(name));
+    if (template) {
+      // Remove previously rendered messages to clear for new message.
+      const messageElement = this.form_.querySelector('.amp-form-messages');
+      if (messageElement) {
+        removeElement(messageElement);
+      }
+      this.templates_.renderTemplate(template, data).then(element => {
+        element.classList.add('amp-form-messages');
+        this.form_.appendChild(element);
```
Comment (erwinmombay) [inline extensions/amp-form/0.1/amp-form.js]: can you use a document fragment here

File: extensions/amp-form/0.1/amp-form.js [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+  }
+
+  /** @private */
+  installSubmitHandler_() {
+    this.form_.addEventListener('submit', e => {
+      // Safari does not trigger validation check on submission, hence we
+      // trigger it manually. In other browsers this would never execute since
+      // the submit event wouldn't be fired if the form is invalid.
+      if (!this.form_.checkValidity()) {
+        this.form_.classList.remove('valid');
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: Ditto on class names.

Why "valid"/"invalid" class names? Why not `:invalid` and `:valid`?

File: extensions/amp-form/0.1/amp-form.js [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        return false;
+      }
+
+      this.form_.classList.add('valid');
+      if (this.form_.classList.contains('invalid')) {
+        this.form_.classList.remove('invalid');
+        this.form_.classList.add('was-invalid');
+      }
+
+      this.xhr_.fetchJson(this.action_, {
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: Of course, except for this one.

File: extensions/amp-form/0.1/amp-form.js [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+  }
+
+  /**
+   * @param {!Object} response.
+   * @private
+   */
+  successHandler_(response) {
+    this.form_.classList.remove('error');
+    this.form_.classList.add('success');
+    this.renderTemplate_('success', response);
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: This is ok for a first PR. But see my comments on the doc about this.

File: extensions/amp-form/0.1/amp-form.js [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+  }
+
+  /**
+   * @param {!Object} error.
+   * @private
+   */
+  errorHandler_(error) {
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: Should be probably refactored into a single method with arguments

File: extensions/amp-form/0.1/amp-form.js [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+   * @param {string} name
+   * @param {!Object} data
+   * @private
+   */
+  renderTemplate_(name, data) {
+    const template = childElement(this.form_,
+        el => el.tagName == 'TEMPLATE' && el.hasAttribute(name));
+    if (template) {
+      // Remove previously rendered messages to clear for new message.
+      const messageElement = this.form_.querySelector('.amp-form-messages');
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: I don't think this is necessary. `[success]`/`[error]` is enough.

File: extensions/amp-form/0.1/amp-form.js [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+function installSubmissionHandlers(win) {
+  onDocumentReady(win.document, () => {
+    toArray(win.document.forms).forEach(form => {
+      new AmpForm(form, win);
+    });
+  });
+}
+
+export function installAmpForms(win) {
+  return getService(win, 'amp-forms', () => {
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: Let's change to `amp-form`. Or whichever way, but we must always install the service with the name of the extension. We can have two services if we want to keep `amp-forms` name.

File: extensions/amp-form/0.1/amp-form.js [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+    installSubmissionHandlers(win);
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: I'd just move the experiment check here.

File: extensions/amp-form/0.1/amp-form.js [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+        element.classList.add('amp-form-messages');
+        this.form_.appendChild(element);
+      });
+    }
+  }
+
+}
+
+function installSubmissionHandlers(win) {
+  onDocumentReady(win.document, () => {
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: I think this is totally fine for now.

File: extensions/amp-form/0.1/amp-form.js [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+    toArray(win.document.forms).forEach(form => {
+      new AmpForm(form, win);
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: What would prevent these instances from garbage collected? Event listeners?

File: src/dom.js [hunk 21]
```diff
@@ -328,13 +329,7 @@ export function childElementsByAttr(parent, attr) {
...
   }
   if (scopeSelectorSupported) {
-    const nodeList = parent.querySelectorAll(':scope > [' + attr + ']');
-    // Convert NodeList into Array.<Element>.
-    const children = [];
-    for (let i = 0; i < nodeList.length; i++) {
-      children.push(nodeList[i]);
-    }
-    return children;
+    return toArray(parent.querySelectorAll(':scope > [' + attr + ']'));
```
Comment (dvoytenko) [inline src/dom.js]: Ideally we'd do this in a separate PR.

File: src/service/xhr-impl.js [hunk 22]
```diff
@@ -22,7 +22,7 @@ import {
   parseQueryString,
   parseUrl,
 } from '../url';
-import {isArray, isObject} from '../types';
+import {isArray, isObject, isFormData} from '../types';
```
Comment (dvoytenko) [inline src/service/xhr-impl.js]: Ditto: separate PR.

File: src/service/xhr-impl.js [hunk 23]
```diff
@@ -223,7 +223,7 @@ function setupJson_(init) {
   init.headers = init.headers || {};
   init.headers['Accept'] = 'application/json';
 
-  if (init.method == 'POST') {
+  if (init.method == 'POST' && init.type == 'json') {
```
Comment (dvoytenko) [inline src/service/xhr-impl.js]: What is `init.type`? Is it part of the spec? I've never seen it. Could you please send me a reference to the doc?

File: src/dom.js (hunk 21 above)
Comment (mkhatib) [inline src/dom.js]: Sent a separate PR #3389
